core/sr-arithmetic/fuzzer/src/biguint.rs

```diff
@@ -0,0 +1,208 @@
+//! More information about `honggfuzz` can be found
+//! [here](https://docs.rs/honggfuzz/0.5.45/honggfuzz/).
+
+use honggfuzz::fuzz;
+use sr_arithmetic::biguint::{BigUint, Single};
+use std::convert::TryFrom;
+use num_traits::ops::checked::CheckedDiv;
+
+type S = u128;
+
+fn main() {
+	loop {
+		fuzz!(|data: (Vec<Single>, Vec<Single>, bool)| {
+			let (mut digits_u, mut digits_v, rem) = data;
+
+			run_with_data_set(4, &digits_u, &digits_v, |u, v| {
+				let ue = S::try_from(u.clone()).unwrap();
+				let ve = S::try_from(v.clone()).unwrap();
+				assert_eq!(u.cmp(&v), ue.cmp(&ve));
+				assert_eq!(u.eq(&v), ue.eq(&ve));
+			});
+
+			run_with_data_set(3, &digits_u, &digits_v, |u, v| {
+				let expected = S::try_from(u.clone()).unwrap() + S::try_from(v.clone()).unwrap();
+				let t = u.clone().add(&v);
+				assert_eq!(
+					S::try_from(t.clone()).unwrap(), expected,
+					"{:?} + {:?} ===> {:?} != {:?}", u, v, t, expected,
+				);
+			});
+
+			run_with_data_set(4, &digits_u, &digits_v, |u, v| {
+				let expected = S::try_from(u.clone())
+					.unwrap()
+					.checked_sub(S::try_from(v.clone()).unwrap());
+				let t = u.clone().sub(&v);
+				if expected.is_none() {
+					assert!(t.is_err())
+				} else {
+					let t = t.unwrap();
+					let expected = expected.unwrap();
+					assert_eq!(
+						S::try_from(t.clone()).unwrap(), expected,
+						"{:?} - {:?} ===> {:?} != {:?}", u, v, t, expected,
+					);
+				}
+			});
+
+			run_with_data_set(2, &digits_u, &digits_v, |u, v| {
+				let expected = S::try_from(u.clone()).unwrap() * S::try_from(v.clone()).unwrap();
+				let t = u.clone().mul(&v);
+				assert_eq!(
+					S::try_from(t.clone()).unwrap(), expected,
+					"{:?} * {:?} ===> {:?} != {:?}", u, v, t, expected,
+				);
+			});
+
+			run_with_data_set(4, &digits_u, &digits_v, |u, v| {
+				let ue = S::try_from(u.clone()).unwrap();
+				let ve = S::try_from(v.clone()).unwrap();
+				if ve == 0 {
+					return;
+				}
+				let (q, r) = (ue / ve, ue % ve);
+				if let Some((qq, rr)) = u.clone().div(&v, true) {
+					assert_eq!(
+						S::try_from(qq.clone()).unwrap(), q,
+						"{:?} / {:?} ===> {:?} != {:?}", u, v, qq, q,
+					);
+					assert_eq!(
+						S::try_from(rr.clone()).unwrap(), r,
+						"{:?} % {:?} ===> {:?} != {:?}", u, v, rr, r,
+					);
+				} else if v.len() == 1 {
+					let qq = u.clone().div_unit(ve as Single);
+					assert_eq!(
+						S::try_from(qq.clone()).unwrap(), q,
+						"[single] {:?} / {:?} ===> {:?} != {:?}", u, v, qq, q,
+					);
+				} else if v.msb() != 0 && u.msb() != 0 && u.len() > v.len() {
+					panic!("div returned none for an unexpected reason");
+				}
+			});
+
+			// Test against num_bigint
+
+			let u = BigUint::from_limbs(&digits_u);
+			let v = BigUint::from_limbs(&digits_v);
+
+			digits_u.reverse();
+			digits_v.reverse();
+
+			let num_u = num_bigint::BigUint::new(digits_u.clone());
+			let num_v = num_bigint::BigUint::new(digits_v.clone());
+
+			// Equality
+
+			assert_biguints_eq(&u, &num_u);
```

Comment:
This function is useful later on when you apply arithmetic, but here it is quite trivial. Both `num_bigint` and `BigUint` are built directly from the same slice. 

What you could do, is to compare the `u.cmp(v)` and `num_u.cmp(num_v)`. Now you are just checking `if digits_u == digits_v`

Reply:
🏴